# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: 6818224b9b11
# Full Hash: 6818224b9b119944f818554e4071a566bca4ed63
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2019-11-05 21:49:23
# Regressor Bug: 1168606
# File Overlap Count: 1
# Description:
#   Bug 1168606 - Add support for preloading key-only cursors. r=ttung,asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44009
# ==============================================================================

diff -r f1be81af7d6b -r 6818224b9b11 dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Tue Nov 05 12:45:48 2019 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Tue Nov 05 12:46:16 2019 +0000
@@ -15217,8 +15217,9 @@
   MOZ_ASSERT_IF(
       aResponse.type() == CursorResponse::Tnsresult ||
           aResponse.type() == CursorResponse::Tvoid_t ||
-          aResponse.type() == CursorResponse::TObjectStoreKeyCursorResponse ||
-          aResponse.type() == CursorResponse::TIndexKeyCursorResponse,
+          aResponse.type() ==
+              CursorResponse::TArrayOfObjectStoreKeyCursorResponse ||
+          aResponse.type() == CursorResponse::TArrayOfIndexKeyCursorResponse,
       aFiles.IsEmpty());
   MOZ_ASSERT(!mActorDestroyed);
   MOZ_ASSERT(mCurrentlyRunningOp);
@@ -25729,7 +25730,12 @@
   Transaction()->AssertIsOnConnectionThread();
   MOZ_ASSERT(aInitializeResponse ==
              (mResponse.type() == CursorResponse::T__None));
-  MOZ_ASSERT_IF(mFiles.IsEmpty(), aInitializeResponse);
+  MOZ_ASSERT_IF(
+      mFiles.IsEmpty() &&
+          (mResponse.type() ==
+               CursorResponse::TArrayOfObjectStoreCursorResponse ||
+           mResponse.type() == CursorResponse::TArrayOfIndexCursorResponse),
+      aInitializeResponse);
 
   nsresult rv = mCursor->mPosition.SetFromStatement(aStmt, 0);
   if (NS_WARN_IF(NS_FAILED(rv))) {
@@ -25774,8 +25780,16 @@
     }
 
     case OpenCursorParams::TObjectStoreOpenKeyCursorParams: {
-      MOZ_ASSERT(aInitializeResponse);
-      mResponse = ObjectStoreKeyCursorResponse(mCursor->mPosition);
+      if (aInitializeResponse) {
+        mResponse = nsTArray<ObjectStoreKeyCursorResponse>();
+      } else {
+        MOZ_ASSERT(mResponse.type() ==
+                   CursorResponse::TArrayOfObjectStoreKeyCursorResponse);
+      }
+
+      auto& responses = mResponse.get_ArrayOfObjectStoreKeyCursorResponse();
+      auto& response = *responses.AppendElement();
+      response.key() = mCursor->mPosition;
       break;
     }
 
@@ -25832,10 +25846,18 @@
         return rv;
       }
 
-      MOZ_ASSERT(aInitializeResponse);
-      mResponse = IndexKeyCursorResponse(mCursor->mPosition,
-                                         mCursor->mLocaleAwarePosition,
-                                         mCursor->mObjectStorePosition);
+      if (aInitializeResponse) {
+        mResponse = nsTArray<IndexKeyCursorResponse>();
+      } else {
+        MOZ_ASSERT(mResponse.type() ==
+                   CursorResponse::TArrayOfIndexKeyCursorResponse);
+      }
+
+      auto& responses = mResponse.get_ArrayOfIndexKeyCursorResponse();
+      auto& response = *responses.AppendElement();
+      response.key() = mCursor->mPosition;
+      response.sortKey() = mCursor->mLocaleAwarePosition;
+      response.objectKey() = mCursor->mObjectStorePosition;
       break;
     }
 
@@ -25851,14 +25873,6 @@
     const nsCString& aOperation) {
   AssertIsOnConnectionThread();
 
-  if (mCursor->mType != OpenCursorParams::TObjectStoreOpenCursorParams &&
-      mCursor->mType != OpenCursorParams::TIndexOpenCursorParams) {
-    IDB_WARNING(
-        "PRELOAD: Not yet implemented. Extra results were queried, but are "
-        "discarded for now.");
-    return NS_OK;
-  }
-
   // For unique cursors, we need to skip records with the same key. The SQL
   // queries currently do not filter these out.
   Key previousKey =